File: electron/main/services/markdown-export.ts
/**
 * Markdown Export Service
 *
 * Converts Claude Code sessions to Markdown format for export
 */

import { format } from 'date-fns'
import type {
  Session,
  ProcessedMessage,
  SubagentSession,
  ToolUseBlock,
  ToolResultBlock,
  ThinkingBlock,
} from '../../shared/session-types'

/**
 * Format a timestamp for display
 */
function formatTimestamp(timestamp: string): string {
  try {
    return format(new Date(timestamp), 'h:mm a')
  } catch {
    return timestamp
  }
}

/**
 * Format full date for header
 */
function formatFullDate(timestamp: number | null): string {
  if (!timestamp) return 'Unknown date'
  try {
    return format(new Date(timestamp), 'MMMM d, yyyy')
  } catch {
    return 'Unknown date'
  }
}

/**
 * Calculate session duration
 */
function formatDuration(startTime: number | null, endTime: number | null): string | null {
  if (!startTime || !endTime) return null
  const durationMs = endTime - startTime
  const minutes = Math.floor(durationMs / 60000)
  const hours = Math.floor(minutes / 60)
  if (hours > 0) {
    return `${hours}h ${minutes % 60}m`
  }
  return `${minutes}m`
}

/**
 * Sanitize a string for use in a filename
 */
export function sanitizeFilename(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, '') // Remove special characters
    .replace(/\s+/g, '-') // Replace spaces with hyphens
    .replace(/-+/g, '-') // Replace multiple hyphens with single
    .replace(/^-|-$/g, '') // Trim hyphens from ends
    .slice(0, 50) // Limit length
}

/**
 * Generate a filename for the export
 * Format: yyyyMMdd_HHmmss_<name>.md
 */
export function generateExportFilename(session: Session): string {
  const timestamp = session.startTime ? new Date(session.startTime) : new Date()
  const dateStr = format(timestamp, 'yyyyMMdd_HHmmss')

  // Use first message as name, or fallback to session id
  const firstUserMessage = session.messages.find((m) => m.role === 'user')
  const name = firstUserMessage
    ? sanitizeFilename(firstUserMessage.textContent.slice(0, 50))
    : session.id.slice(0, 8)

  return `${dateStr}_${name || 'session'}.md`
}

/**
 * Format tool use block for markdown
 */
function formatToolUse(tool: ToolUseBlock, result: ToolResultBlock | undefined): string {
  const lines: string[] = []

  lines.push(`<details>`)
  lines.push(`<summary><strong>Tool:</strong> ${tool.name}</summary>`)
  lines.push('')

  // Tool input
  if (Object.keys(tool.input).length > 0) {
    lines.push('**Input:**')
    lines.push('```json')
    lines.push(JSON.stringify(tool.input, null, 2))
    lines.push('```')
  }

  // Tool result
  if (result) {
    lines.push('')
    lines.push(result.is_error ? '**Error:**' : '**Result:**')
    const content = typeof result.content === 'string' ? result.content : JSON.stringify(result.content)
    // Truncate very long results
    const truncated = content.length > 2000 ? content.slice(0, 2000) + '\n... (truncated)' : content
    lines.push('```')
    lines.push(truncated)
    lines.push('```')
  }

  lines.push('</details>')
  lines.push('')

  return lines.join('\n')
}

/**
 * Format thinking block for markdown
 */
function formatThinking(thinking: ThinkingBlock): string {
  const lines: string[] = []
  lines.push('<details>')
  lines.push('<summary><em>Thinking...</em></summary>')
  lines.push('')
  lines.push(thinking.thinking)
  lines.push('')
  lines.push('</details>')
  lines.push('')
  return lines.join('\n')
}

/**
 * Format a single message for markdown
 */
function formatMessage(message: ProcessedMessage, subagents: Record<string, SubagentSession>): string {
  const lines: string[] = []
  const timestamp = formatTimestamp(message.timestamp)
  const roleLabel = message.role === 'user' ? '**User**' : '**Assistant**'

  lines.push(`### ${roleLabel} (${timestamp})`)
  lines.push('')

  // Thinking blocks (for assistant messages)
  if (message.thinkingBlocks.length > 0) {
    for (const thinking of message.thinkingBlocks) {
      lines.push(formatThinking(thinking))
    }
  }

  // Main text content
  if (message.textContent) {
    lines.push(message.textContent)
    lines.push('')
  }

  // Tool use blocks (for assistant messages)
  if (message.toolUseBlocks.length > 0) {
    for (const tool of message.toolUseBlocks) {
      const result = message.toolResults[tool.id]

      // Check if this is a Task tool with a subagent
      if (tool.name === 'Task' && tool.agentId && subagents[tool.agentId]) {
        lines.push(formatSubagent(subagents[tool.agentId]))
      } else {
        lines.push(formatToolUse(tool, result))
      }
    }
  }

  return lines.join('\n')
}

/**
 * Format a subagent session for markdown
 */
function formatSubagent(subagent: SubagentSession): string {
  const lines: string[] = []

  lines.push('<details>')
  lines.push(`<summary><strong>Subagent</strong> (${subagent.messageCount} messages)</summary>`)
  lines.push('')

  for (const message of subagent.messages) {
    // Format subagent messages with simpler formatting (no nested subagents)
    const timestamp = formatTimestamp(message.timestamp)
    const roleLabel = message.role === 'user' ? '**User**' : '**Assistant**'

    lines.push(`#### ${roleLabel} (${timestamp})`)
    lines.push('')

    if (message.textContent) {
      lines.push(message.textContent)
      lines.push('')
    }

    // Tool uses in subagent
    for (const tool of message.toolUseBlocks) {
      const result = message.toolResults[tool.id]
      lines.push(formatToolUse(tool, result))
    }
  }

  lines.push('</details>')
  lines.push('')

  return lines.join('\n')
}

/**
 * Convert a session to Markdown format
 */
export function sessionToMarkdown(session: Session): string {
  const lines: string[] = []

  // Header
  lines.push(`# Session: ${session.project}`)
  lines.push('')

  // Metadata
  lines.push(`**Date:** ${formatFullDate(session.startTime)}`)

  const duration = formatDuration(session.startTime, session.endTime)
  if (duration) {
    lines.push(`**Duration:** ${duration}`)
  }

  lines.push(`**Messages:** ${session.messages.length}`)

  if (session.gitBranch) {
    lines.push(`**Branch:** ${session.gitBranch}`)
  }

  if (session.version) {
    lines.push(`**Claude Code Version:** ${session.version}`)
  }

  lines.push(`**Session ID:** ${session.id}`)
  lines.push('')
  lines.push('---')
  lines.push('')
  lines.push('## Conversation')
  lines.push('')

  // Messages
  for (const message of session.messages) {
    lines.push(formatMessage(message, session.subagents))
  }

  // Footer
  lines.push('---')
  lines.push('')
  lines.push(`*Exported from Sessionly on ${format(new Date(), 'MMMM d, yyyy \'at\' h:mm a')}*`)

  return lines.join('\n')
}

File: src-tauri/src/markdown_export.rs
use crate::session_types::{Session, ProcessedMessage, SubagentSession, ToolUseBlock, ToolResultBlock, ThinkingBlock};
use chrono::{DateTime, Utc, TimeZone};
use std::collections::HashMap;

fn format_timestamp(timestamp: &str) -> String {
    if let Ok(dt) = DateTime::parse_from_rfc3339(timestamp) {
        dt.format("%l:%M %p").to_string().trim().to_string()
    } else {
        timestamp.to_string()
    }
}

fn format_full_date(timestamp: Option<i64>) -> String {
    match timestamp {
        Some(ms) => {
            if let Some(dt) = Utc.timestamp_millis_opt(ms).single() {
                dt.format("%B %e, %Y").to_string()
            } else {
                "Unknown date".to_string()
            }
        }
        None => "Unknown date".to_string(),
    }
}

fn format_duration(start: Option<i64>, end: Option<i64>) -> Option<String> {
    match (start, end) {
        (Some(s), Some(e)) => {
            let minutes = ((e - s) / 60000) as u64;
            let hours = minutes / 60;
            if hours > 0 {
                Some(format!("{}h {}m", hours, minutes % 60))
            } else {
                Some(format!("{}m", minutes))
            }
        }
        _ => None,
    }
}

fn format_tool_use(tool: &ToolUseBlock, result: Option<&ToolResultBlock>) -> String {
    let mut lines = Vec::new();
    lines.push("<details>".to_string());
    lines.push(format!("<summary><strong>Tool:</strong> {}</summary>", tool.name));
    lines.push(String::new());

    if tool.input != serde_json::Value::Object(Default::default()) {
        lines.push("**Input:**".to_string());
        lines.push("```json".to_string());
        lines.push(serde_json::to_string_pretty(&tool.input).unwrap_or_default());
        lines.push("```".to_string());
    }

    if let Some(result) = result {
        lines.push(String::new());
        lines.push(if result.is_error == Some(true) { "**Error:**" } else { "**Result:**" }.to_string());
        let content = if let Some(s) = result.content.as_str() {
            s.to_string()
        } else {
            serde_json::to_string(&result.content).unwrap_or_default()
        };
        let truncated = if content.len() > 2000 {
            format!("{}\n... (truncated)", &content[..2000])
        } else {
            content
        };
        lines.push("```".to_string());
        lines.push(truncated);
        lines.push("```".to_string());
    }

    lines.push("</details>".to_string());
    lines.push(String::new());
    lines.join("\n")
}

fn format_thinking(thinking: &ThinkingBlock) -> String {
    format!(
        "<details>\n<summary><em>Thinking...</em></summary>\n\n{}\n\n</details>\n",
        thinking.thinking
    )
}

fn format_subagent(subagent: &SubagentSession) -> String {
    let mut lines = Vec::new();
    lines.push("<details>".to_string());
    lines.push(format!("<summary><strong>Subagent</strong> ({} messages)</summary>", subagent.message_count));
    lines.push(String::new());

    for message in &subagent.messages {
        let ts = format_timestamp(&message.timestamp);
        let role = if message.role == "user" { "**User**" } else { "**Assistant**" };
        lines.push(format!("#### {} ({})", role, ts));
        lines.push(String::new());
        if !message.text_content.is_empty() {
            lines.push(message.text_content.clone());
            lines.push(String::new());
        }
        for tool in &message.tool_use_blocks {
            let result = message.tool_results.get(&tool.id);
            lines.push(format_tool_use(tool, result));
        }
    }

    lines.push("</details>".to_string());
    lines.push(String::new());
    lines.join("\n")
}

fn format_message(message: &ProcessedMessage, subagents: &HashMap<String, SubagentSession>) -> String {
    let mut lines = Vec::new();
    let ts = format_timestamp(&message.timestamp);
    let role = if message.role == "user" { "**User**" } else { "**Assistant**" };
    lines.push(format!("### {} ({})", role, ts));
    lines.push(String::new());

    for thinking in &message.thinking_blocks {
        lines.push(format_thinking(thinking));
    }

    if !message.text_content.is_empty() {
        lines.push(message.text_content.clone());
        lines.push(String::new());
    }

    for tool in &message.tool_use_blocks {
        let result = message.tool_results.get(&tool.id);
        if tool.name == "Task" {
            if let Some(agent_id) = &tool.agent_id {
                if let Some(subagent) = subagents.get(agent_id) {
                    lines.push(format_subagent(subagent));
                    continue;
                }
            }
        }
        lines.push(format_tool_use(tool, result));
    }

    lines.join("\n")
}

pub fn session_to_markdown(session: &Session) -> String {
    let mut lines = Vec::new();

    lines.push(format!("# Session: {}", session.project));
    lines.push(String::new());
    lines.push(format!("**Date:** {}", format_full_date(session.start_time)));

    if let Some(dur) = format_duration(session.start_time, session.end_time) {
        lines.push(format!("**Duration:** {}", dur));
    }

    lines.push(format!("**Messages:** {}", session.messages.len()));

    if let Some(branch) = &session.git_branch {
        lines.push(format!("**Branch:** {}", branch));
    }

    if !session.version.is_empty() {
        lines.push(format!("**Claude Code Version:** {}", session.version));
    }

    lines.push(format!("**Session ID:** {}", session.id));
    lines.push(String::new());
    lines.push("---".to_string());
    lines.push(String::new());
    lines.push("## Conversation".to_string());
    lines.push(String::new());

    for message in &session.messages {
        lines.push(format_message(message, &session.subagents));
    }

    lines.push("---".to_string());
    lines.push(String::new());

    let now = Utc::now().format("%B %e, %Y at %l:%M %p");
    lines.push(format!("*Exported from Sessionly on {}*", now));

    lines.join("\n")
}
